x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java

```diff
@@ -0,0 +1,202 @@
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class Engine implements Writeable, ToXContentObject {
+    private final String name;
+    private final String[] indices;
+
+    /**
+     * Public constructor.
+     *
+     * @param name The name of the engine.
+     * @param indices The list of indices targeted by this engine.
+     */
+    public Engine(String name, String[] indices) {
+        this.name = name;
+        this.indices = indices;
+    }
+
+    public Engine(StreamInput in) throws IOException {
+        this.name = in.readString();
+        this.indices = in.readStringArray();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(name);
+        out.writeStringArray(indices);
+    }
+
+    private static final ConstructingObjectParser<Engine, Void> PARSER = new ConstructingObjectParser<>("engine", false, params -> {
+        final String name = (String) params[0];
+        @SuppressWarnings("unchecked")
+        final String[] indices = ((List<String>) params[1]).toArray(String[]::new);
+        return new Engine(name, indices);
+    });
+    private static final ParseField NAME_FIELD = new ParseField("name");
+    private static final ParseField INDICES_FIELD = new ParseField("indices");
+
+    static {
+        PARSER.declareString(constructorArg(), NAME_FIELD);
+        PARSER.declareStringArray(constructorArg(), INDICES_FIELD);
+    }
+
+    /**
+     * Parses a {@link Engine} from its {@param xContentType} representation in bytes.
+     *
+     * @param resourceName The name of the resource (must match the {@link Engine} name).
+     * @param source The bytes that represents the {@link Engine}.
+     * @param xContentType The format of the representation.
+     *
+     * @return The parsed {@link Engine}.
+     */
+    public static Engine fromXContentBytes(String resourceName, BytesReference source, XContentType xContentType) {
+        try (XContentParser parser = XContentHelper.createParser(XContentParserConfiguration.EMPTY, source, xContentType)) {
+            return Engine.fromXContent(resourceName, parser);
+        } catch (IOException e) {
+            throw new ElasticsearchParseException("Failed to parse the get result", e);
```

Comment:
Do we have an opportunity for a clearer Exception message here?

Reply:
I added the source that failed to parse in the message. Is it what you had in mind?